feature/win-end-screen

FellonOne opened the pull request:
add win/end screen;
add several level on one game;
fix issues in __levels;

FellonOne committed:
add win/loose screen;
add change level functional;
fix issues in __levels;
fix issues with render;

FellonOne committed:
fix issues in sagas on Game store;
add clearLevel after complete level;

art4vm commented on app/src/store/Game/reducer.ts:
```diff
@@ -20,6 +21,12 @@ export const gameReducer = (state = gameState, action: GameActions): GameState =
+            newState.countLevels = action.payload;
+
+            return newState;
+        }
```
В этом нет необходимости, можно ставить ко-во уровней в SET_LEVELS_ORDER.
При выполнении SET_LEVELS_ORDER уже устанавливается массив уровней и он **всегда** будет равен countLevels. А еще лучше убрать счетчик из этого хранилища и добавить его в хранилище levels(обновляется при добавлении уровней) тк это значение всегда будет одинаковым для всех игр

art4vm commented on app/src/store/Game/reducer.ts:
```diff
@@ -34,7 +41,7 @@ export const gameReducer = (state = gameState, action: GameActions): GameState =
         case GAME_ACTION_TYPES.SET_LEVEL_NUMBER: {
             const newState = cloneDeep(state);
-            if (action.payload) {
+            if (action.payload !== undefined) {
```
в этом не необходимости

FellonOne replied:
есть, когда меняешть на значение 0, оно не попадает в ветвь и не меняется

art4vm replied:
Предполагается, что явно передавать уровень не нужно, все ставится автоматом в sagas в стадии interlude

art4vm commented on app/src/store/Game/sagas.ts:
```diff
@@ -32,9 +38,14 @@ function* gameSetStateWorker(action: GameSetStateAction) {
+                const shuffleArray = shuffle(Array.from({length: levels.length}, () => c--)).slice(
+                    0,
+                    Math.floor(Math.random() * 5) + 1
+                );
```
зачем рандомизировать длину игры? В таком случае нет смысла считать очки, тк они будут примерно пропорциональны кол-ву уровней

FellonOne replied:
в теории да(
Обсудим вечером тогда. Можно зафиксировать на 3-4 уровня за игру

kdeit reviewed (changes requested):
Есть какой-то баг.
Начал уровень - Пишет 1/4
Попался призраку - Try Again - Пишет 1/3
Прошёл - Пишет 1/2
Потом попался — Опять 1/3

FellonOne commented:
> Есть какой-то баг.
> Начал уровень - Пишет 1/4
> Попался призраку - Try Again - Пишет 1/3
> Прошёл - Пишет 1/2
> Потом попался — Опять 1/3

Это все из-за рандомизации кол-ва уровней

FellonOne committed:
add fixed level counts;
fix key location in __levels;
update win screen;

art4vm committed:
Небольшие правки игры

Удалено лишнее из gameStore
Немного поправлены вызовы в engine
Поправлены стартовые координаты в __levels

art4vm commented:
В рамках своего ПР с leaderboard сделаю еще правки

art4vm commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Смена уровня](https://trello.com/c/Kfi0Z2gn/86-%D1%81%D0%BC%D0%B5%D0%BD%D0%B0-%D1%83%D1%80%D0%BE%D0%B2%D0%BD%D1%8F)
